orttraining/orttraining/test/graph/graph_parser_test.cc

```diff
@@ -0,0 +1,1256 @@
+  embed_layer_norm_node->SetExecutionProviderType(layer_norm_node.GetExecutionProviderType());
+}
+
+TEST(GraphParser, base1) {
+  PatternGraph g(
+      {Constant<double>("C"),
+       Constant<double>("X")},
+      {PNode("Add", {"C", "X"}, {"Y"}, "CX-Y"),
+       PNode("Exp", {"Y"}, {"Z"}, "Y-Z")},
+      "", "test_graph");
+
+  Graph& graph = g.GetGraph();
+  GraphViewer graph_viewer(graph);
+  const auto& node_topology_list = graph_viewer.GetNodesInTopologicalOrder();
+  for (auto node_index : node_topology_list) {
+    auto* node = graph.GetNode(node_index);
+    std::cout << "Name: " << node->Name() << ", OpType: " << node->OpType()
+              << ", Domain: " << node->Domain() << ", SinceVersion: " << node->SinceVersion() << std::endl;
+  }
+}
+
+TEST(GraphParser, base2) {
+  PatternGraph g(
+      {Constant<double>("C1"),
+       Constant<double>("C2"),
+       Constant<double>("X")},
+      {PNode("Exp", {"X"}, {"Y"}, "Exp"),
+       PNode("Add", {"Y", "C1"}, {"Z"}, "Add"),
+       PNode("Sub", {"Z", "C2"}, {"W"}, "Sub")},
+      "", "pattern_graph");
+
+  Graph& graph = g.GetGraph();
+  GraphViewer graph_viewer(graph);
+  const auto& node_topology_list = graph_viewer.GetNodesInTopologicalOrder();
+  for (auto node_index : node_topology_list) {
+    auto* node = graph.GetNode(node_index);
+    std::cout << "Name: " << node->Name() << ", OpType: " << node->OpType()
+              << ", Domain: " << node->Domain() << ", SinceVersion: " << node->SinceVersion() << std::endl;
+  }
+}
+
+TEST(GraphParser, match1) {
+  PatternGraph target(
+      {Constant<double>("C1"),
+       Constant<double>("C2"),
+       Constant<double>("X")},
+      {PNode("Exp", {"X"}, {"Y"}, "Exp"),
+       PNode("Add", {"Y", "C1"}, {"Z1"}, "gAdd"),
+       PNode("Add", {"Y", "C2"}, {"Z2"}, "customize"),
+       PNode("Sub", {"Z1", "Z2"}, {"W"}, "Sub")},
+      "", "target_graph");
+  PatternGraph pattern(
+      {Constant<double>("C1"),
+       Constant<double>("C2"),
+       Constant<double>("X")},
+      {PNode("Exp", {"X"}, {"Y"}, "Exp"),
+       PNode("Add", {"Y", "C1"}, {"Z"}, "customize"),
+       PNode("Sub", {"Z", "C2"}, {"W"}, "Sub")},
+      "Exp", "pattern_graph");
+
+  class CustomNodeCompareFunc : public NodeCompareFunc {
+   public:
+    bool operator()(const Node* g, const PNode* p, const Graph& /*target*/,
+                    const PatternGraph& /*pattern*/) const override {
+      return p->NameEquals(g->Name());
+    }
+  };
+
+  pattern.SetCustomConstraint(std::make_unique<CustomNodeCompareFunc>());
+  Graph& target_graph = target.GetGraph();
+
+  std::vector<PNN> res;
+  ASSERT_TRUE(pattern.TryMatch(target_graph, res).IsOK());
+
+  for (auto node_pair : res) {
+    auto g = target_graph.GetNode(node_pair.first);
+    auto p = node_pair.second;
+    std::cout << "Graph node: " << g->Name() << "   Pattern node: " << p->Name() << std::endl;
+  }
+}
+
+TEST(GraphParser, match2) {
+  PatternGraph target(
+      {Constant<double>("X"),
```

Comment:
>Constant [](http://example.com/codeflow?start=7&length=8)

as we talked earlier, we need make most of inputs as initializers, which can be created with a function differently with Constant.

Reply:
I use ```IArg``` instead of ```Constant``` to support the feature. Parameter ```is_constant``` could be specified by user to tell that whether an arg is constant.